### PR TITLE
DMT Phase 2 Doc Updates

### PR DESCRIPTION
**Pull request by rebekah-lawrence**
After the conversations following the previous release, I have restructured the DMT section.

Instead of getting everyone to review this restructure AND the v2 updates, I have separated out this stage to ensure we end up with useful, useable information once the updates are included

Feel free to correct any incorrect/unclear information, but for this round, please concentrate on the structure.

### CODE REVIEW

**Review by frant-hartm — CHANGES_REQUESTED** [review]
This is a huge improvement, thanks!

File: docs/modules/migrate/pages/dmt-connection-config.adoc
Comment (rebekah-lawrence) [reply docs/modules/migrate/pages/dmt-connection-config.adoc:75]: The `:` on a line generally communicates further content that are/might be present in the code; it has been standard doc practice for a long time

However, if it s going to cause confusion, we'll need a different approach. We could: 

* Include the omitted lines (which may also muddy the water of what is actually being shown)
* Only include the changed lines with no additional context (I'd recommend keeping context if possible)
* Use a different notation (ideas welcome)
* Explicitly state at the start that lines containing `:` identify places where other lines might exist

File: docs/modules/migrate/pages/dmt-migration.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-migration.adoc:49]:
```suggestion
* Example connection configuration file, which is used to connect to the migration cluster
```

File: docs/modules/migrate/pages/dmt-migration.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-migration.adoc:51]: no longer true we removed this

File: docs/modules/migrate/pages/dmt-view-details.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-view-details.adoc:10]: I don't know these details by my heart. I think @yuce should review this PR as well to be sure that frontend related stuff. If nothing changed from the old content though, we should be fine. cc @rebekah-lawrence

File: docs/modules/migrate/pages/dmt-run.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-run.adoc:53]: I think it's not in this PR but we need to update the command reference. --config flag for dmt migration cluster config to connect to migration cluster is missing, also --yes flag etc.(not sure if we should add it but good to do )

File: docs/modules/migrate/pages/dmt-connection-config.adoc
Comment (rebekah-lawrence) [reply docs/modules/migrate/pages/dmt-connection-config.adoc:75]: After discussion with the team, standard notation will be three periods (`...`) on the line instead of the colon (`:`)

File: docs/modules/migrate/pages/dmt-view-details.adoc
Comment (rebekah-lawrence) [reply docs/modules/migrate/pages/dmt-view-details.adoc:10]: Invited Yuce to review; this will form part of the updates for v2 work (which I am starting now)

File: docs/modules/migrate/pages/dmt-run.adoc
Comment (rebekah-lawrence) [reply docs/modules/migrate/pages/dmt-run.adoc:53]: I didn't do the Command Reference last time, so it didn't occur to me!

@srknzl do you want me to pick this up, or would you prefer that the team continues to maintain this page?

**Review by frant-hartm — APPROVED** [review]

File: docs/modules/migrate/pages/dmt-start-migration-cluster.adoc
Comment (yuce) [inline docs/modules/migrate/pages/dmt-start-migration-cluster.adoc:23]: This section may be updated in case named CLC configuration is to be used.

File: docs/modules/migrate/pages/dmt-start-migration-cluster.adoc
Comment (yuce) [inline docs/modules/migrate/pages/dmt-start-migration-cluster.adoc:27]: Is this necessary?

File: docs/modules/migrate/pages/dmt-run.adoc
Comment (yuce) [reply docs/modules/migrate/pages/dmt-run.adoc:52]: IMO it would be the best if the user created a named configuration using `clc config add` instead of providing the configuration path directly. In that case they don't need to provide the configuration with each command. CLC will display a configuration selector if necessary.

File: docs/modules/migrate/pages/dmt-view-data.adoc
Comment (yuce) [inline docs/modules/migrate/pages/dmt-view-data.adoc:30]: The new output is:
```
------
 Size 
------
 1000 
------
    OK Returned 1 row(s).
```

**Review by yuce — CHANGES_REQUESTED** [review]
Great update, thanks Rebekah!
Just a few ideas...

File: docs/modules/migrate/pages/dmt-run.adoc
Comment (srknzl) [reply docs/modules/migrate/pages/dmt-run.adoc:52]: > IMO it would be the best if the user created a named configuration using clc config add instead of providing the configuration path directly. In that case they don't need to provide the configuration with each command. CLC will display a configuration selector if necessary.

yeah that'll be very good. What if they don't have clc though? In any case, I think we should have --config flag in the command reference because it's available for use and I don't see a point in hiding it. 

@rebekah-lawrence @yuce and I can do that, thank you

Comment (srknzl) [reply docs/modules/migrate/pages/dmt-run.adoc:52]: @rebekah-lawrence note: one possible update I see is that adding a note about what @yuce said, basically they can create a named clc config so that they don't have to specify "migration.json" config path all the time via --config flag. 

@yuce is it possible to call "dmt config add"?

File: docs/modules/migrate/pages/dmt-start-migration-cluster.adoc
Comment (rebekah-lawrence) [reply docs/modules/migrate/pages/dmt-start-migration-cluster.adoc:27]: Debatable ;-)

I chose to include it for the following reasons:

* It is mentioned as a step in the High-level Process section, so it ought to be reflected in the How-to topics for completeness and to reduce cognitive load for people while reading
* When people use docs, they might not be thinking as they follow the steps and could inadvertently miss a step that would otherwise be in their mind
* It was raised by Hendy when she commented on the original single-page version of this information

Bottom line is, it is short enough, and clear enough, that people can easily skip it if they don't need the reminder/full picture; but, it provides enough support to cover the first two reasons given above for including it

File: docs/modules/migrate/pages/dmt-run.adoc
Comment (rebekah-lawrence) [reply docs/modules/migrate/pages/dmt-run.adoc:52]: I have added a note, called it `dmt config add` for now along with a x-ref to the DMT Command Reference assuming the `dmt config` command will be added there. Can easily change if this isn't implemented as dmt-specific.

File: docs/modules/migrate/pages/dmt-start-migration-cluster.adoc
Comment (rebekah-lawrence) [reply docs/modules/migrate/pages/dmt-start-migration-cluster.adoc:23]: Not changed for the moment, will revisit when approach is confirmed

File: docs/modules/migrate/pages/dmt-run.adoc
Comment (yuce) [reply docs/modules/migrate/pages/dmt-run.adoc:53]: @srknzl 
DMT has the same `config` commands with CLC.

Comment (srknzl) [reply docs/modules/migrate/pages/dmt-run.adoc:53]: Rebekah gave a link to dmt command reference. We should update DMT command reference and state that clc commands like "clc config" are also usable with dmt

Comment (srknzl) [reply docs/modules/migrate/pages/dmt-run.adoc:53]: https://github.com/hazelcast/hz-docs/pull/981

File: docs/modules/migrate/pages/data-migration-tool.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/data-migration-tool.adoc:20]: We'll add an option to enable/disable replicated map metadata migration via a flag, will let you know when that happens

File: docs/modules/migrate/pages/dmt-connection-config.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-connection-config.adoc:53]:
```suggestion
To configure the migration of CP data structures, you must update the _config.json_ file located in the _migration_config_ folder.
```

we'll not have an option about cp structures I think.  so we can move this sentence to a different place and say:

To configure the data migration, you must update the _config.json_ file located in the _migration_config_ folder.

Comment (srknzl) [reply docs/modules/migrate/pages/dmt-connection-config.adoc:53]: It might be reasonable to keep this PR open until we finish the config.json addition then. Not to have to open a new pr

File: docs/modules/migrate/pages/dmt-migration.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-migration.adoc:2]: for now, the cp configuration is only txt files where you list their names. we can try to allow them not requiring a cluster restart if they give us a list of names instead of a wildcard. then there will be a config.json option for that. 

but as I know, we'll not do it. am i right? @frant-hartm

Comment (srknzl) [reply docs/modules/migrate/pages/dmt-migration.adoc:2]: this means for now, we can delete "if configured" part above. if there wont be a extra config for cp structures in config.json

File: docs/modules/migrate/pages/dmt-run.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-run.adoc:65]: yes they can restart the migration. currently the way to do that is just restarting the migration with the same command. existing data structures in the target cluster will be skipped.

File: docs/modules/migrate/pages/dmt-connection-config.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-connection-config.adoc:41]: For ReplicatedMap with metadata, we should add a note. "It's not possible to migrate replicated map metadata to target versions below 5.3.7." This kind of limitation does not exist for CP. They can be migrated to a target cluster with version 5.2.z for instance.

Comment (srknzl) [reply docs/modules/migrate/pages/dmt-connection-config.adoc:41]: Another thing, we should clarify that the custom jar is for the source cluster and not the target cluster. Also we can give a link to the new page I added in this PR https://github.com/hazelcast/hz-docs/pull/1042

File: docs/modules/migrate/pages/data-migration-tool.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/data-migration-tool.adoc:22]: almost true. 

If the user does not set the "migrateReplicatedMapMetadata" option to false explicitly, the migration cluster will try to migrate replicated map metadata by default and there will be an error & migration fail about the missing API.

The CP migration is explicit by the user, we only migrate CP migrations if there are entries in the CP files in  migration_config/data folder, used to select which CP structures to migrate. If they select CP data structures but the API is absent on the source side, we'll again fail the migration.

File: docs/modules/migrate/pages/data-migration-tool.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/data-migration-tool.adoc:30]:
```suggestion
** The target cluster must be running on one of the following:
```

File: docs/modules/migrate/pages/data-migration-tool.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/data-migration-tool.adoc:32]: 5.3.2 is for the IMap metadata migration, (which can't be disabled by an option), not for ReplicatedMap

File: docs/modules/migrate/pages/data-migration-tool.adoc
Comment (rebekah-lawrence) [reply docs/modules/migrate/pages/data-migration-tool.adoc]: Great spot, thank you!
Amended in [https://github.com/hazelcast/hz-docs/pull/936/commits/58c391dd98438c1981eac72bf13fe637ebb02c3d](url)

File: docs/modules/migrate/pages/data-migration-tool.adoc
Comment (rebekah-lawrence) [reply docs/modules/migrate/pages/data-migration-tool.adoc]: I have amended and expanded this description in the latest update.

File: docs/modules/migrate/pages/data-migration-tool.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/data-migration-tool.adoc:28]: If you have not added the extension JAR or disabled the migration of Replicated map metadata, the migration fails with an error about the missing API contained in the extension JAR.

--> 

If you have not added the extension JAR and have not disabled the migration of Replicated map metadata, the migration fails with an error about the missing API contained in the extension JAR.

File: docs/modules/migrate/pages/data-migration-tool.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/data-migration-tool.adoc:30]: since we say if needed, lets change here as well. an extension jar is needed if they want to migrate cp structures. 

So, "if needed" means they selected cp structures to migrate.

if they selected cp structures to migrate, but haven't added the extension jar to the classpath the behaviour is not "CP data structure are not migrated."

it's failing the migration with an error about the missing CP APIs that are in the extension jar or 5.3.7+ cluster version.

File: docs/modules/migrate/pages/dmt-dev-test.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-dev-test.adoc:22]: do we still use the word IMDG? Is it because the page is only available on 4.2? Is there a platform version of it? newer, for 5.x

File: docs/modules/migrate/pages/dmt-migration.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-migration.adoc:6]:
```suggestion
NOTE: To migrate Replicated metadata and/or CP structures, you must include the provided JAR in your class path. The source cluster must be restarted after the addition of the JAR.
```

Comment (srknzl) [reply docs/modules/migrate/pages/dmt-migration.adoc:6]: Also, maybe remove this note as it may be confusing.


or we can give a link to the related place without removing it.

File: docs/modules/migrate/pages/dmt-connection-config.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-connection-config.adoc:34]: also there are new files for cp structures. see https://github.com/hazelcast/hazelcast-mono/tree/0778d11cc407c3677c3a166308aecf5c01a5c418/distribution/src/root/migration_config/data

Comment (srknzl) [reply docs/modules/migrate/pages/dmt-connection-config.adoc:34]: I would just remove the file names from here as there are 5 of them now

File: docs/modules/migrate/pages/dmt-connection-config.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-connection-config.adoc:77]:
```suggestion
|Defines whether to include data structures with names that start with a double underscore ($$__$$) in the migration when a wildcard (`*`) is included in the `run` command. For example, Hazelcast uses some internal maps with names starting with $$__$$ such as $$__$$sql.catalog.
```

This option is a generic option that can be used with any data structure migration. If there are other places, with this description, we should also update them

Comment (srknzl) [reply docs/modules/migrate/pages/dmt-connection-config.adoc:77]: Also, I believe we should mention "migration config" instead of "run" command to be more clear with our description.

Comment (srknzl) [reply docs/modules/migrate/pages/dmt-connection-config.adoc:77]: Plus, since this option is generic, I don't think we should mention in the CP structures section, maybe in "Update Data Structures" section above.

File: docs/modules/migrate/pages/dmt-view-data.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-view-data.adoc:2]: with the addition of CP migration support, we may want to update this page with more generic "data structure" migration validation. it can be either

- map
- replicated map 
- cp structures

for map and replicated map, size check should be enough. for cp structures, checking the existence of such a cp structure with the provided name should be good enough

Comment (srknzl) [reply docs/modules/migrate/pages/dmt-view-data.adoc:2]: As I know, clc does not support replicated map, atomic reference, count down latch.

We can direct users to management center for those.

File: docs/modules/migrate/pages/dmt-connection-config.adoc
Comment (srknzl) [reply docs/modules/migrate/pages/dmt-connection-config.adoc:77]: For reference, this is my description of the config option in the dmt package: https://github.com/hazelcast/hazelcast-mono/blob/d12b08f726de9dff17f2e06ac5188e9f52109d0b/distribution/src/root/migration_config/migration_config.yaml#L3 

I added this description.

**Review by srknzl — COMMENTED** [review]
looks good in general, however still can be improved imo. ty for the changes in advance

**Review by frant-hartm — APPROVED** [review]

File: docs/modules/migrate/pages/dmt-dev-test.adoc
Comment (rebekah-lawrence) [reply docs/modules/migrate/pages/dmt-dev-test.adoc:22]: The example was to show the migration path - people can substitute their product/version and still follow the steps. I chose an older version because (a) it seems to be a common migration use case and (b) to highlight the steps required to update multiple versions.

As you say, it is fine, so I won't amend this.

File: docs/modules/migrate/pages/dmt-view-data.adoc
Comment (srknzl) [inline docs/modules/migrate/pages/dmt-view-data.adoc:19]: not sure if this is correct. "--map" should be for the imaps. let me confirm it with Yuce

Comment (srknzl) [reply docs/modules/migrate/pages/dmt-view-data.adoc:19]: confirmed that clc does not support replicated map

Comment (rebekah-lawrence) [reply docs/modules/migrate/pages/dmt-view-data.adoc]: [e3b6c3b](https://github.com/hazelcast/hz-docs/pull/936/commits/e3b6c3b17d71b48118a3cf9c9e2e607c215ed112)

File: docs/modules/migrate/pages/dmt-view-data.adoc
Comment (rebekah-lawrence) [reply docs/modules/migrate/pages/dmt-view-data.adoc]: [e3b6c3b](https://github.com/hazelcast/hz-docs/pull/936/commits/e3b6c3b17d71b48118a3cf9c9e2e607c215ed112)

**Review by srknzl — APPROVED** [review]

**Comment by srknzl** [comment]
Sorry I realized now. CLC has atomicreference command now it seems: https://github.com/hazelcast/clc/pull/69

Can we update the related "verify your data migration" part accordingly? @rebekah-lawrence

**Comment by srknzl** [comment]
@yuce what's the proper way of checking if a atomic reference exists using clc? `clc atomicreference is-null` ? However, if an atomic reference does not exist, we don't want to create it via this command. Is it available to do that?

**Comment by yuce** [comment]
@srknzl Any `atomicreference` operation would create a proxy if it doesn't already exist, just like the client libraries.
I would assume creating the proxy also creates an atomic reference on the server-side.
Currently, there's no way to check whether an atomic reference exists on the server, using CLC.
To my knowledge, it's not possible to easily do that without CLC as well.

File: docs/modules/migrate/pages/data-migration-tool.adoc
Comment (sumnerib) [inline docs/modules/migrate/pages/data-migration-tool.adoc:18]: Maybe an additional disclaimer around this part that mentions that these timings are just examples, and that individual migration times will depend on various factors including environmental constraints (i.e. network, hardware resources, etc.). Otherwise LGTM.

**Review by sumnerib — CHANGES_REQUESTED** [review]

**Comment by srknzl** [comment]
> @srknzl Any atomicreference operation would create a proxy if it doesn't already exist, just like the client libraries.
I would assume creating the proxy also creates an atomic reference on the server-side.
Currently, there's no way to check whether an atomic reference exists on the server, using CLC.
To my knowledge, it's not possible to easily do that without CLC as well.

Hmm, as far as I recall, there was a note in our documentation (or somewhere in confluence) about CP structures not being created until some value is written to them. So suppose if I do: "return getAtomicReference("foo").get() == null". Will this create atomic reference on the cluster @arodionov ?

**Comment by arodionov** [comment]
Yes, the proxy on the client side will not automatically create the object on the server side.
So,
        `IAtomicReference<String> ref = cpSubsystem.getAtomicReference("iRef");`
will not create a server-side object.
But 
        `String val = ref.get();`
will create it with a `null` value.

**Comment by arodionov** [comment]
Yes, the source code for this operation is the following:
```
        V atomicValue = atomicValues.get(key);
        if (atomicValue == null) {
            atomicValue = newAtomicValue(groupId, name, null);
            atomicValues.put(key, atomicValue);
        }
        return atomicValue;
```

**Comment by rebekah-lawrence** [comment]
I have updated the 5.3.7 requirement to 5.3.8 - @frant-hartm could you take another final look and make sure I have correctly understood

**Review by srknzl — APPROVED** [review]
the version change is lgtm

**Comment by rebekah-lawrence** [comment]
Adding a comment to ensure this is not automatically Closed before it is confirmed that this can be merged.

**Pull request closed without merging**
